text/0000-00-00-coprocessor-write.md

```diff
@@ -0,0 +1,154 @@
+# Coprocessor Write
+
+## Summary
+
+This RFC intends to extend the ability of coprocessor to support writing. The user can send a _single_ coprocessor request to do a TiDB query (a.k.a. a DAG request) first and then perform a write request (like prewrite or pessimistic lock) with the result of the previous TiDB query.
+
+## Motivation
+
+Currently, all write requests are initiated from a TiKV client. To update some data, the user typically reads from TiKV first and prepares the data to write. And then, it sends a write request to TiKV. The whole process includes at least two RPCs between the client and TiKV. There is usually extra cost of encoding and decoding.
+
+However, in multiple cases, we can move some of the logic from the client to TiKV to reduce one RPC and save encoding/decoding cost.
+
+### Simple updates
+
+If the result set of a DAG request is the same as the data to write, we can let coprocessor invoke prewrite.
+
+Suppose we have a typical table with a primary key and secondary indexes:
+
+| Column | Type | Key             |
+| ------ | ---- | --------------- |
+| id     | INT  | primary key     |
+| v      | INT  |                 |
+| idx    | INT  | secondary index |
+
+If we update the non-index column `v` according to the primary key `id`, the rows we query are right the rows we are going to write.
+
+For example, if we do `UPDATE t SET v = v + 1 WHERE id = 1`, we can query TiKV with `SELECT id, v + 1, idx FROM t WHERE id = 1`. Then, the result row is identical to the row we want to write. It will be easy for coprocessor to also do a prewrite and return the prewrite result to the client.
+
+### Pessimistic lock rows with conditions
+
+In pessimistic transactions, keys to be written need to be locked before the 2PC process. Only mutations with known row handle can be locked with a single RPC. In other cases, we need two RPCs: the first one to get the rows and the second one to lock the rows.
+
+Using the same table in the previous section, when we want to do `UPDATE t SET idx = idx + 1 WHERE v BETWEEN 10 AND 20`, we do a TiDB query with `SELECT * FROM t WHERE v BETWEEN 10 AND 20` first, and then we lock the rows with `AcquirePessimisticLock` RPCs.
+
+We can find the result set contains sufficient information to construct the row keys we want to lock. It would be better if we can use only one RPC to do the query and lock these keys. Coprocessor can take the result set of the query as the input of `AcquirePessimisticLock`. This can save one RPC.
+
+## Detailed design
```

Comment:
We could add some execution flow to explain how the optimized scenarios will work. From the SQL execution perspective.